core/src/main/java/kafka/server/SharePartitionManager.java

```diff
@@ -157,13 +157,18 @@ private void maybeProcessFetchQueue() {
                 // Add the share partition to the list of partitions to be fetched only if we can
                 // acquire the fetch lock on it.
                 if (sharePartition.maybeAcquireFetchLock()) {
-                    if (sharePartition.canAcquireMore()) {
-                        topicPartitionData.put(topicIdPartition, new FetchRequest.PartitionData(
-                                topicIdPartition.topicId(),
-                                sharePartition.nextFetchOffset(),
-                                0,
-                                partitionMaxBytes,
-                                Optional.empty()));
+                    // Fetching over a topic should be able to proceed if any one of the following 2 conditions are met:
+                    // 1. The fetch is to happen somewhere in between the record states cached in the share partition.
+                    //    This is because in this case we don't need to check for the partition limit for in flight messages
+                    // 2. If condition 1 is not true, then that means we will be fetching new records which haven't been cached before.
+                    //    In this case it is necessary to check if the partition limit for in flight messages has been reached.
+                    if (sharePartition.nextFetchOffset() != sharePartition.endOffset() || sharePartition.canAcquireMore()) {
```

Comment:
I think this is along the right lines, but I'm not sure it's quite right.

If I understand correctly, `endOffset` is the SPEO and it points to the final record which has already been fetched, while `nextFetchOffset` would point to the next offset to fetch. So, I wonder whether this is essentially out by one. I might be confused, but I don't think this is quite achieving exactly what condition (1) requires.

Reply:
Thanks for reviewing the PR. You are absolutely right, nextFetchOffset should point to endOffset + 1. I wonder why the scenario did not produce an error locally in this case. Anyways, I will push the changes in the next commit.